### ISSUE TITLE
[Arquillian Cube OpenShift] namespace.use.existing + env.init.enabled=true fails to create test resources

### ISSUE THREAD

**Issue by RadekKoubsky** [arquillian/arquillian-cube#743]
##### Issue Overview

I have my existing openshift project which si empty. I want arquillian cube to deploy test resources (e.g. openshift.yml) generated by FMP during resource goal. When running integration tests, arquillian cube throws the following exception about non-existing image stream:

```
Running io.openshift.booster.OpenShiftIT
Initializing Session:b13b197f-9d14-49f0-a628-571ccd39cef7
Using Kubernetes at: https://api.devel.xpaas:8443/
Did not find any kubernetes configuration.
Applying additional kubernetes configuration from: file:/home/rkoubsky/QE/MSA/forks/rkoubsky/rest_springboot-tomcat/target/spring-boot-http-is.yml
Could not annotate namespace: [rkoubsky] with status: [ERROR].
Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 3.949 sec <<< FAILURE! - in io.openshift.booster.OpenShiftIT
io.openshift.booster.OpenShiftIT  Time elapsed: 3.948 sec  <<< ERROR!
java.lang.RuntimeException: io.fabric8.kubernetes.clnt.v2_2.KubernetesClientException: Failure executing: POST at: https://api.devel.xpaas:8443/oapi/v1/namespaces/rkoubsky/imagestreams. Message: ImageStream "spring-boot-http" is invalid: []: Internal error: imagestreams "spring-boot-http" is invalid: spec.tags[latest].from.name: Invalid value: "spring-boot-http@sha256:ae65653b28535fac376fc62f9b4fda24206152964ca1b6884c621cc2e633698d": error generating tag event: imagestreamimage "sha256:ae65653b28535fac376fc62f9b4fda24206152964ca1b6884c621cc2e633698d" not found. Received status: Status(apiVersion=v1, code=422, details=StatusDetails(causes=[StatusCause(field=[], message=Internal error: imagestreams "spring-boot-http" is invalid: spec.tags[latest].from.name: Invalid value: "spring-boot-http@sha256:ae65653b28535fac376fc62f9b4fda24206152964ca1b6884c621cc2e633698d": error generating tag event: imagestreamimage "sha256:ae65653b28535fac376fc62f9b4fda24206152964ca1b6884c621cc2e633698d" not found, reason=InternalError, additionalProperties={})], group=null, kind=ImageStream, name=spring-boot-http, retryAfterSeconds=null, additionalProperties={}), kind=Status, message=ImageStream "spring-boot-http" is invalid: []: Internal error: imagestreams "spring-boot-http" is invalid: spec.tags[latest].from.name: Invalid value: "spring-boot-http@sha256:ae65653b28535fac376fc62f9b4fda24206152964ca1b6884c621cc2e633698d": error generating tag event: imagestreamimage "sha256:ae65653b28535fac376fc62f9b4fda24206152964ca1b6884c621cc2e633698d" not found, metadata=ListMeta(resourceVersion=null, selfLink=null, additionalProperties={}), reason=Invalid, status=Failure, additionalProperties={}).
Caused by: io.fabric8.kubernetes.clnt.v2_2.KubernetesClientException: Failure executing: POST at: https://api.devel.xpaas:8443/oapi/v1/namespaces/rkoubsky/imagestreams. Message: ImageStream "spring-boot-http" is invalid: []: Internal error: imagestreams "spring-boot-http" is invalid: spec.tags[latest].from.name: Invalid value: "spring-boot-http@sha256:ae65653b28535fac376fc62f9b4fda24206152964ca1b6884c621cc2e633698d": error generating tag event: imagestreamimage "sha256:ae65653b28535fac376fc62f9b4fda24206152964ca1b6884c621cc2e633698d" not found. Received status: Status(apiVersion=v1, code=422, details=StatusDetails(causes=[StatusCause(field=[], message=Internal error: imagestreams "spring-boot-http" is invalid: spec.tags[latest].from.name: Invalid value: "spring-boot-http@sha256:ae65653b28535fac376fc62f9b4fda24206152964ca1b6884c621cc2e633698d": error generating tag event: imagestreamimage "sha256:ae65653b28535fac376fc62f9b4fda24206152964ca1b6884c621cc2e633698d" not found, reason=InternalError, additionalProperties={})], group=null, kind=ImageStream, name=spring-boot-http, retryAfterSeconds=null, additionalProperties={}), kind=Status, message=ImageStream "spring-boot-http" is invalid: []: Internal error: imagestreams "spring-boot-http" is invalid: spec.tags[latest].from.name: Invalid value: "spring-boot-http@sha256:ae65653b28535fac376fc62f9b4fda24206152964ca1b6884c621cc2e633698d": error generating tag event: imagestreamimage "sha256:ae65653b28535fac376fc62f9b4fda24206152964ca1b6884c621cc2e633698d" not found, metadata=ListMeta(resourceVersion=null, selfLink=null, additionalProperties={}), reason=Invalid, status=Failure, additionalProperties={}).


Results :

Tests in error: 
  OpenShiftIT.io.openshift.booster.OpenShiftIT » Runtime io.fabric8.kubernetes.c...

Tests run: 1, Failures: 0, Errors: 1, Skipped: 0

```

When I run:
```
mvn clean fabric8:deploy -Popenshift
mvn verify -Popenshift-it -Dnamespace.use.existing=my-project -Denv.init.enabled=false
``` 
It works fine.

##### Expected Behaviour
Integration test deploys its resources and test them

##### Current Behaviour
Integration test fails to deploy its resources.

##### Steps To Reproduce
1. clone [this maven project](https://github.com/RadekKoubsky/rest_springboot-tomcat/tree/arquillian-cube-integration-test-openshift)
2. Checkout branch `arquillian-cube-integration-test-openshift`
3. run `mvn clean verify -Popenshift,openshift-it -Dnamespace.use.existing=my-project -Denv.init.enabled=true`

**Comment by RadekKoubsky** [comment]
I have tried the code with Arquillian Cube 1.8.0, the issue with image stream is still present.

**Comment by lordofthejars** [comment]
@RadekKoubsky I have started working on this. Any thought where the error could happen? I am asking just to not going completely blind.

**Comment by lordofthejars** [comment]
@RadekKoubsky I have been reading the issue and running the example and I have noticed that I am not sure if an error on Cube or if you have another scenario in mind but look when you do:

```
mvn clean fabric8:deploy -Popenshift
mvn verify -Popenshift-it -Dnamespace.use.existing=my-project -Denv.init.enabled=false
```

It works. And it works because `fabric8:deploy` generates the kubernetes resource and deploy it to OpenShift and then test is run, and since everything is deployed it works.

Then you say that using just thins command:

```
mvn clean verify -Popenshift,openshift-it -Dnamespace.use.existing=my-project -Denv.init.enabled=true
```
Does not work. Basically it does not work because Arquillian is not deploying the scenario, but here is my question. There is no really advantage of using first instead of second approach. In terms of CI/CD it doesn't care since it is Jenkinsfile who does everything. In terms of development in terminal, well he can create a shell script and run it from build tool (Maven), create an alias or whatever. Then there is if developer is running from IDE, in this case is where it is interesting that Arquillian could deploy the services but you still need the manual step of generating the kubernetes file, so again if you need a manual step, it is not so automatic at all.

The only thing I see is that you can generate kubernetes file with maven call then switch to IDE and run the test from there, but then not sure it is valid from UX.

Am I missing something?

**Comment by Ladicek** [comment]
@lordofthejars The correct one to speak to is @spisiakm, as @RadekKoubsky is leaving Red Hat.

Actually the 1st scenario, the one with 2 Maven invocations, is something we don't want. We want the second one to work, the one with a single Maven invocation.

**Comment by lordofthejars** [comment]
Ok great, welcome @spisiakm, but what is the benefit of the second one if you still need one manual step?

**Comment by Ladicek** [comment]
Sorry, not sure I'm following. Here's a somewhat more recent description of the issue: https://github.com/openshiftio/appdev-planning/issues/43#issuecomment-318640464

Basically, we need `mvn clean verify` to work, in an existing namespace. Are we doing something wrong?

**Comment by lordofthejars** [comment]
but why not `mvn fabric8:deploy verify -D.....` ?

**Comment by Ladicek** [comment]
Arquillian Cube is supposed to deploy the application, no? I'm pretty sure I'm missing something.

**Comment by lordofthejars** [comment]
Yes when it is necessary (basically when you want to run your application from IDE or from CLI), basically when you create manually your `kubernetes.json` then it has sense to do it. But in this application you are not creating any `kubernetes.json` manually, you are letting fabric8 plugin to do it, so it means that you need to run at least one time (until you `clean` the project) the maven goal to generate these resources.

What I am saying is that if you want to run your tests from Maven, you don't need that Arquillian deploys anything you can just run `mvn fabric8:deploy verify -D.....` and that's all. Why would you need Arquillian does anything?

The other thing which might be more useful might be an integration between fabric8 plugin and arquillian so you can run safely from IDE and from CLI without worrying about calling the plugin first.

But this is a big change overall and I would like to listen comments about it from @bartoszmajsak and @iocanel and @jwendell

**Comment by lordofthejars** [comment]
@Ladicek if you want we can move the conversation to BJN as well.

**Comment by Ladicek** [comment]
So, in the linked repo, if I do `mvn clean verify -Popenshift,openshift-it -Dnamespace.use.existing=myproject -Denv.init.enabled=true`, the `fabric8:resource` and `fabric8:build` goals run and the `{kubernetes,openshift}.{json,yml}` files are created long before the test starts. So I expect Arquillian to take them, apply them, run the test, then cleanup.

You ask: why do you want to do this? Well I ask: why not? It's simpler, it's how Maven is typically used, and if I understand the docs correctly, it should work. And it seems it would, but the image stream created by `fabric8:build` gets deleted for some reason. Is Arquillian possibly doing some pre-test cleanup?

**Comment by Ladicek** [comment]
Oh thanks for the BJN offer, but I'm not very familiar with this issue, so unless you are fine with making it a pair debugging session, it wouldn't work :-)

**Comment by lordofthejars** [comment]
Exactly we can do this thing you mention but you could do `mvn clean fabric8:deploy verify -Popenshift,openshift-it -Dnamespace.use.existing=myproject -Denv.init.enabled=false` and you'll get the same, what I mean is why do you want that Arquillian deploys something generated by Maven plugin when the Maven plugin itself can do it.

**Comment by Ladicek** [comment]
Well, at the very least, I don't want to force users to add more commands to the cmdline, when the traditional `mvn clean verify` obviously should work.

But out of curiosity, I tried `mvn clean fabric8:deploy verify -Popenshift,openshift-it -Dnamespace.use.existing=myproject -Denv.init.enabled=false` on the linked repo, and `fabric8:deploy` doesn't even get a chance to run, because tests are executed first.

**Comment by lordofthejars** [comment]
humm ok then tomorrow morning I will take a look what you should do to make it work and provide you the CLI option to do it.

**Comment by Ladicek** [comment]
Thank you!

**Comment by bartoszmajsak** [comment]
We have discussed potential solutions for this functionality with @lordofthejars.

### Quick win

To let you guys use one mvn command to run everything we will fix Cube to scan for files generated by `fabric8-maven-plugin`. @lordofthejars is on it.

### Desired improvement

We have two options here if we consider use case when you run the test from the IDE. We need to have those ` {kubernetes,openshift}.{json,yml}` files generated upfront.

  * we can re-use logic of `fabric8-maven-plugin` to
    * generate those files
    * create s2i image and register it in openshift registry
    * this can be done under the hood based on e.g. `@AutogenerateOpenshiftResources` annotation put on the test class
  * we can provide `Shrinkwrap` descriptors so you can either create default files like fmp does but also open it up for customization "shrinkwrap style"
    * this would mean we have to flesh out what kind of `@Deployment` methods we would like to have to make it simple and powerful enough

Thoughts?

**Comment by Ladicek** [comment]
Re "Quick win": my understanding from [this section](http://arquillian.org/arquillian-cube/#_how_to_run_kubernetes_with_multiple_configuration_files) of the documentation is that Arquillian Cube should already be able to find the FMP-generated files. Am I wrong here? I'd consider this a bugfix, not an improvement. Actually you're using the word "fix" too, so it seems we're on the same page :-)

Re "Desired improvement": I don't think we considered running tests from IDE just yet :-) I think all our Kubernetes/OpenShift tooling revolves around FMP, so the 1st option sounds best to me. But I don't have all the details, I might be easily missing something.

And thanks for paying attention to this, it's gonna be a huge help for us!

**Comment by bartoszmajsak** [comment]
Ok, let's make the "Quick win" working then and we can discuss IDE experience when you will be focusing on it. I think we have a bug here which @lordofthejars is looking at.

> And thanks for paying attention to this, it's gonna be a huge help for us!

Sure, our pleasure :)

**Comment by lordofthejars** [comment]
We have found the root cause. Currently fabric8 plugin generate resources at `target/classes/META-INF/fabric8` classpath location. The problem is that `surefire` plugin does not add `target/classes` but `target/test-classes` (`"surefire.test.class.path" -> "/Users/alex/git/rest_springboot-tomcat/target/test-classes:/Users/alex/git/rest_springboot-tomcat/target/spring-boot-http-7-SNAPSHOT.jar:/Users/alex/.m2/repository/org/apache/tomcat/tomcat-juli/8.0.36/tomcat-juli-8.0.36.jar:..."` so for Arquillian is impossible to find these files.

What we are going to do is talk with fabric8 guys to see if the plugin can copy resources in both places.

**Comment by Ladicek** [comment]
That sounds weird. Per http://maven.apache.org/surefire/maven-surefire-plugin/examples/configuring-classpath.html, the `target/classes` directory is on test classpath. Is that maybe a Surefire/Failsafe bug? Let me check.

**Comment by lordofthejars** [comment]
Currently you are using `failsafe`

**Comment by Ladicek** [comment]
Failsafe documentation is the same in this regard: http://maven.apache.org/components/surefire/maven-failsafe-plugin/examples/configuring-classpath.html

**Comment by bartoszmajsak** [comment]
Documentation is apparently wrong. In the surefire classpath what you are getting is a jar file with your production code, but not `target/classes`

**Comment by Ladicek** [comment]
Ah, I see! That's again the thing with Failsafe preferring to use the JAR instead of `target/classes`!

**Comment by Ladicek** [comment]
OK, so if I downgrade to Failsafe 2.18.1, which still uses `target/classes`, I get further. The test fails when trying to apply `target/spring-boot-http-is.yml` with an error message of:

```
imagestreams "spring-boot-http" is invalid: spec.tags[latest].from.name: Invalid value: "spring-boot-http@sha256:cf768ac1152974a20a08a34c76066a97137a1f0c29c1d57215dad183b9215b3b
```

This image _did_ exist before the test started, as it was created by the Fabric8 Maven plugin. I can see it in the OpenShift web console. I don't know who deleted it -- was it Arquillian Cube?

(This finding is consistent with the issue description in https://github.com/openshiftio/appdev-planning/issues/43#issuecomment-318640464)

**Comment by lordofthejars** [comment]
The problem here is that fabric8 executes this `-is.yml` file, and then in cube we are rexecuting again and you get this failure. In fact if you use CLI you'll get the same error. Figuring out how to fix this.

**Comment by Ladicek** [comment]
I think you can easily apply the same file twice. The issue seems to be that FMP creates the image, the image stream and the image stream tag, then someone deletes it all, and then Cube tries to apply the `-is.yml` file, which tries to create the image stream and the image stream tag, but the image is missing.

Anyway, great we're on the same page! :-)

**Comment by lordofthejars** [comment]
exactly this is what is happening, now I have commented the code in cube that detects `-is.yml` files and it has been able to deploy everything. So this is a really good step forward, so I need to think a bit but maybe we will make this `-is.yml` detection code optional being default to false.

cc/ @bartoszmajsak @dipak-pawar

**Comment by Ladicek** [comment]
Great! I'm just wondering, who and why deletes the image+imagestream+imagestreamtag? That shouldn't happen, should it?

**Comment by Ladicek** [comment]
Also, instead of downgrading to Failsafe 2.18.1, adding this to the Failsafe plugin configuration seems to work:

```xml
<configuration>
    <additionalClasspathElements>
        <additionalClasspathElement>${project.build.outputDirectory}/META-INF/fabric8</additionalClasspathElement>
    </additionalClasspathElements>
</configuration>
```

**Comment by lordofthejars** [comment]
Good news I have got a running version of the project. I will add a new flag, and release 1.9.0 version of cube which by default will ignore the `-is.yml` files.

**Comment by Ladicek** [comment]
Cool! Any idea why the image gets deleted and applying the `-is.yml` file subsequently fails? I still don't quite understand why that happens.

**Comment by lordofthejars** [comment]
I think that it is not really delated at all, for what we have seen is that is trying to create resources twice. Look if you want to try by yourself do this. `mvn clean package` and then go to `target` directory and do `oc create -f ...-is.yml` and you'll get the same error.

**Comment by Ladicek** [comment]
Well you're right, `oc create -f ...-is.yml` indeed also fails with `error generating tag event: imagestreamimage.image.openshift.io "sha256:469b67311d56220795dec3c7a39ef2d9f61cd9d2cc25cca99752fb1cee2a0ebb" not found`. That "not found" bit fooled me.

**Comment by lordofthejars** [comment]
yes I think that it is more how OpenShift treat this scenario more than something related to our stuff.

**Comment by Ladicek** [comment]
So the reason why `META-INF/fabric8` stuff couldn't be found on test classpath is this line: https://github.com/snowdrop/spring-boot-http-booster/blob/master/pom.xml#L133 Once removed, it works just fine. (Where just fine = obviously I hit the issue with `-is.yml`.)

**Comment by Ladicek** [comment]
I sent PRs to Spring Boot boosters to avoid binding FMP to the `package` phase. That should solve our issues with finding `META-INF/fabric8` on the test classpath, without actually changing FMP.

- https://github.com/snowdrop/spring-boot-http-booster/pull/58
- https://github.com/snowdrop/spring-boot-http-secured-booster/pull/23
- https://github.com/snowdrop/spring-boot-crud-booster/pull/15
- https://github.com/snowdrop/spring-boot-configmap-booster/pull/30
- https://github.com/snowdrop/spring-boot-health-check-booster/pull/14

**Comment by lordofthejars** [comment]
Ok I have updated kubernetes client on cube to latest ones, also I have fixed the problem with `-is.yml` files. Tomorrow I will release a new version so you can work on Monday.

**Comment by Ladicek** [comment]
Thanks @lordofthejars! Oh and I'm totally fine with release on Monday, we all deserve a weekend off of work!

**Comment by lordofthejars** [comment]
The problem is that Monday is PTO in my case, then Tue and Wed I am at JavaZone so this leave us to Thursday.

**Comment by bartoszmajsak** [comment]
Well I can do computers too, so I can take care of the release.

**Comment by lordofthejars** [comment]
When this is released @Ladicek if you see that it works, could you just ping us so we can close this issue?

**Comment by bartoszmajsak** [comment]
Instead of releasing and trying it out only then maybe you could build snapshot against latest `master` and confirm if that works for you @Ladicek?

**Comment by lordofthejars** [comment]
I tried locally and it worked

**Comment by Ladicek** [comment]
I just tried locally as well and it works. Unless it breaks your schedules or something, I'd like to check a couple more boosters before proceeding with a release. I'll comment later today or tomorrow.

**Comment by bartoszmajsak** [comment]
Thanks @Ladicek. Please run more booster tests and let us know so we gonna release being sure it works for you.

**Comment by spisiakm** [comment]
Thanks a lot @bartoszmajsak @lordofthejars for looking into this and also @Ladicek for substituting me in this case, unfortunately I wasn't available last week. I'm going try it out and see what happens. @Ladicek has already tried this out with Spring Boot http booster and it worked just fine, so thank you again, you helped us quite a bit ! :)

**Comment by bartoszmajsak** [comment]
Just give us go/no-go and we will proceed accordingly :)

**Comment by Ladicek** [comment]
Unfortunately, this isn't enough yet :-( It works flawlessly when running tests in an existing project (which is our main usecase), but running in an Arquillian-created project doesn't work (which, obviously, is needed for OpenShift.io).

1. `mvn clean verify -Popenshift,openshift-it -Dnamespace.use.existing=myproject -Denv.init.enabled=true` works
2. `mvn clean verify -Popenshift,openshift-it -Denv.init.enabled=true` doesn't work, the build gets stuck at `Applying kubernetes configuration from: jar:file:/home/lthon/tmp/rest_springboot-tomcat/target/spring-boot-http-7-SNAPSHOT.jar!/META-INF/fabric8/openshift.json`, no pod is running in the `itest-xxxxx` namespace, presumably because the image was build in another namespace?
3. `mvn clean verify -Popenshift,openshift-it -Denv.init.enabled=true -DenableImageStreamDetection=true` doesn't work, the test fails with `java.lang.RuntimeException: io.fabric8.kubernetes.clnt.v2_6.KubernetesClientException: Failure executing: POST at: https://192.168.0.100:8443/oapi/v1/namespaces/itest-6785a55b/imagestreams. Message: ImageStream "spring-boot-http" is invalid`, which seems to be the same error message this all started with.

Sounds like we're back to square 1?

**Comment by lordofthejars** [comment]
Case 3 is a wrong configuration. Fabric8 automatically creates image
stream. So if you set cube to create image stream it fails because IS is
already created.

Is the same case as if you run oc create .... After running fabric8

**Comment by Ladicek** [comment]
Indeed Fabric8 Maven plugin creates the image stream, but in a different namespace than the `itest-xxxxx`. So I was thinking I need to enable adding the image stream to the `itest-xxxx` namespace, and that seems to work manually.

**Comment by lordofthejars** [comment]
Hummm then I will need to take a look

**Comment by Ladicek** [comment]
BTW, here's what I did to verify the behavior in the 3rd case (all with a snapshot build of current Cube master):

```
mvn clean verify -Popenshift,openshift-it -Denv.init.enabled=true -DenableImageStreamDetection=true -Dnamespace.cleanup.enabled=false -Dnamespace.destroy.enabled=false
# this fails with java.lang.RuntimeException: io.fabric8.kubernetes.clnt.v2_6.KubernetesClientException: Failure executing: POST at: https://192.168.0.100:8443/oapi/v1/namespaces/itest-4d3cff0b/imagestreams. Message: ImageStream "spring-boot-http" is invalid

oc get project
oc project itest-xxxxx

oc create -f target/spring-boot-http-is.yml
# this works just fine, and actually allows the deployment created by Cube before the test to proceed
```

**Comment by Ladicek** [comment]
Also, I've been successful with porting a WildFly Swarm booster test to Arquillian Cube (found https://github.com/arquillian/arquillian-cube/pull/803 in the process), but I stumbled upon this roadblock: https://github.com/Ladicek/wfswarm-rest-http/blob/arquillian-cube/src/test/java/io/openshift/booster/GreetingServiceTest.java#L37 Do you possibly know of any solution? Thanks!

**Comment by lordofthejars** [comment]
Hi @Ladicek about the second and third case that it is not working, this is related on how it works Cube and fabric8. What Fabric8 Maven plugin does is create and configure some stuff inside the default OpenShift namespace (ie `myproject`). And this occurs before Cube test is executed. Then Cube test is executed and it creates a random namespace (`it-....`) but this new namespace has not the content generated by fabric8 maven plugin previously, so when the `-is.yml` is executed it fails because it misses some content.

The solution for that requires some big development, basically an integration between Fabric8 and Cube like explained here https://github.com/arquillian/arquillian-cube/issues/743#issuecomment-327757345 so fabric8 bits are executed after the random namespace is created by cube.

**Comment by lordofthejars** [comment]
@Ladicek I don't understand exactly the problem, but if the problem is just enable tests in failsafe and not surefire, the only thing I guess you can do is playing with inclusions/exclusions in surefire plugin and failsafe plugin cc/ @bartoszmajsak

**Comment by lordofthejars** [comment]
https://github.com/arquillian/arquillian-cube/issues/806

**Comment by Ladicek** [comment]
Hi @lordofthejars, when it comes to the `-is.yml` thing, I think this should easily work, because the `-is.yml` contains a namespace, so when importing it to `itest-xxxx`, the image should be found just fine. And indeed when trying it manually, it works -- as described in https://github.com/arquillian/arquillian-cube/issues/743#issuecomment-328752684

When it comes to the second problem (when Cube is used in a Surefire test run) -- ah you are actually right, I can easily exclude Cube from test classpath with Surefire! THANKS A LOT! Funny I didn't think of it myself, as I've been dealing with test classpath exclusions just a while ago :-)

**Comment by lordofthejars** [comment]
Humm I tried with @dipak-pawar and it failed using `oc`. I'll continue investigating on this too, because in my case it failed as well.

**Comment by Ladicek** [comment]
Hmm weird. I'll check again.

**Comment by lordofthejars** [comment]
I have talked with @rhuss today and it seems that this use case should not be working by default if you are not pointing fabric8 and cube into same namespace.

**Comment by Ladicek** [comment]
So basically the ability of Cube to create a separate namespace is known to not work when using Fabric8 Maven plugin? That's unfortunate.

**Comment by lordofthejars** [comment]
It is not exactly like that I mean it is normal that this is happening since until now I am not sure that Cube was thought to be work together with fabric8 how we are doing now. Currently Fabric8 has its own bits and Cube their own. This means that Fabric8 does some stuff when it is run with the current namespace. On the other side Cube does some stuff on configured namespace which can be the current one or a new one. If you choose a new one, since a) Fabric8 runs before Cube and b) the namespace is generated randomly, we have no way to make it work with random stuff.

Tomorrow I'll try if there is any workaround, but as I suspect we will need to implement this integration so first Cube creates the namespace and second Fabric8 does its stuff and not the other way around.

**Comment by Ladicek** [comment]
Thanks for info, @lordofthejars!

**Comment by Ladicek** [comment]
I'm thinking we should probably open a new issue for the 2nd usecase (FMP + Arquillian Cube creating a new namespace), now that the 1st one works fine (FMP + Arquillian Cube running in current namespace). WDYT?

**Comment by lordofthejars** [comment]
@Ladicek this is exactly what I wanted to talk with you on mattermost. The first thing is to try to define a priority, currently case 1 is covered and it is working, then case 2 and 3 are something that is currently widely used or something that is planed to be used (just to see if it is also a blocker, or it is something that can wait one week).

For second case I created a new issue https://github.com/arquillian/arquillian-cube/issues/806 before starting implementing it, today I'll check if there is some quick workaround.

**Comment by Ladicek** [comment]
@lordofthejars Sorry, I don't frequent Mattermost, it's like a 5th chat I'd have to figure out... but if you ping me there, I can join.

The cases 2/3 -- well we don't need them at the moment, I believe, but it will become important soon because of OpenShift.io. So waiting one or two weeks should be OK.

**Comment by bartoszmajsak** [comment]
@Ladicek can you elaborate a bit on

>  but it will become important soon because of OpenShift.io

So with the fix 1. you can switch from `OpenshiftTestAssistant` to Cube, but those test will still not run on OSIO? https://github.com/openshiftio/booster-common/issues/8

**Comment by Ladicek** [comment]
I keep hearing that we need to switch to Arquillian Cube because it allows running tests in a different namespace, which is needed on OpenShift.io: https://github.com/openshiftio/booster-common/issues/8 I don't know if that's true or not, but assuming it is, then:

1. We can switch to Cube for RHOAR boosters tests, because we run them in "current" namespace. Well I'm pretty sure we'll find issues along the way (one thing I can think of right now: can we deploy a configmap manually before Cube starts deploying resources?), but that's probably expected.
2. We still won't be able to run RHOAR boosters tests on OpenShift.io, because running in a different namespace doesn't work.

**Comment by lordofthejars** [comment]
> configmap manually

yes if you do it either manually before running `mvn` or in maven lifecycle before integration tests.

**Comment by Ladicek** [comment]
That's exactly what I hoped I won't hear :-) But we'll figure it out when we get there.

**Comment by lordofthejars** [comment]
or you can create an openshift additional file and configure cube to pick it up

**Comment by lordofthejars** [comment]
@Ladicek check latest master because the problem described in 3) should be fixed.
Notice that the use case number 2, is not valid here because fabric8 with OpenShift requires an ImageStream insert on given namespace. If you don't set the `enableImageStreamDetection` to true  then Cube never inserts the ImageStream so the test will fail because it misses one part (the Image Stream part).

So I think that now everything is fixed. Just give a try.

**Comment by lordofthejars** [comment]
More about second case notice that in case 2 you are running `mvn clean verify -Popenshift,openshift-it -Denv.init.enabled=true` and in case 3 `mvn clean verify -Popenshift,openshift-it -Denv.init.enabled=true -DenableImageStreamDetection=true` since you want to use `openshift.json` and this file requires an ImageStream there, if you don't set the the flag to true when deploying the openshift.json you'll get an error (as it is happening). So the rule of thumb here is that if you are using OpenShift with ImageStreams (by default in F8 Maven Plugin are used) you need to set the `enableImageStreamDetection ` flag to true or it won't work because of missing pieces on tested namespace.

Of course this do not apply when you are using the same namespace as used in F8 Maven plugin (case 1)

**Comment by lordofthejars** [comment]
@spisiakm Have you had time to try with latest master of Arquillian Cube? In my computer the project using method 3 passed correctly.

**Comment by Ladicek** [comment]
Hi, indeed case 2 is wrong per current design and I expect it to fail. But case 3 definitely should work.

And I just verified that with current Cube master, case 3 works just fine. I think Cube currently covers the usecases we need (though I guess we'll find other issues as we progress :-) ).

@spisiakm, could you please also check and confirm? Thanks.

**Comment by spisiakm** [comment]
Hi folks,
case no. 3 works like a charm now with current master, so I guess this issue is resolved/fixed. Thanks a lot !

**Comment by lordofthejars** [comment]
You are welcome, thank you very much for your feedback, since it is really valuable. Anything you need please ping us, since we are trying to give top priority.

Today I am going to release a new Arquillian Cube version so you have on maven central.

**Comment by Ladicek** [comment]
Thanks @lordofthrjars, having these fixes in an actual release will be great! We realized we'll need to switch to Cube pretty soon, we'll get in touch if we hit an issue. Thanks!

**Comment by bartoszmajsak** [comment]
@Ladicek grab it while it's hot http://search.maven.org/#search|ga|1|arquillian-cube
